MAINT Add floatX function and make ADVI use it.

twiecki opened the pull request:
Alternative to https://github.com/pymc-devs/pymc3/pull/1623

twiecki committed:
MAINT Replace discrete RV check.

MAINT Replace discrete RV check.

springcoil commented:
```ERROR: Failure: ImportError (cannot import name check_discrete_rvs)```

springcoil commented:
I think this is ready to merge - does anyone disagree? @fonnesbeck @twiecki @ferrine @ColCarroll

ferrine commented on pymc3/theanof.py:
```diff
@@ -38,11 +42,16 @@ def cont_inputs(f):
+def floatX(X):
+    return np.asarray(X, dtype=theano.config.floatX)
+
+floatX_str = theano.config.floatX
```
Why do you use this shortcut? It becomes a constant, not reference. So if anyone changes floatX, then this constant will not inherent changes. It can cause unexpected errors.

ferrine commented on pymc3/theanof.py:
```diff
@@ -38,11 +42,16 @@ def cont_inputs(f):
     return typefilter(inputvars(f), continuous_types)
 
 
+def floatX(X):
```
What about `X.astype(...)` to use theano and numpy interface seamlessly?

twiecki committed:
Remove static reference to theano.config.floatX. Use .astype(). Addresses @ferrine's comments.

Remove static reference to theano.config.floatX. Use .astype(). Addresses @ferrine's comments.

twiecki committed:
ENH Add floatX() to theanof and make advi use it.

twiecki committed:
Remove nan_. Add doc-string.

twiecki commented:
@ferrine I pointed this to 3.1 now and addressed your comments.

springcoil commented:
Hmm 

```
======================================================================
ERROR: test_advi_minibatch (pymc3.tests.test_advi.TestADVI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_advi.py", line 179, in test_advi_minibatch
    total_size=n, learning_rate=1e-1)
  File "/home/travis/minconda3/envs/testenv/lib/python3.5/site-packages/theano/configparser.py", line 115, in res
    return f(*args, **kwargs)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/variational/advi_minibatch.py", line 459, in advi_minibatch
    minibatch_tensors, total_size)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/variational/advi_minibatch.py", line 45, in _get_rvss
    s = floatX(total_size) / minibatch_tensors[0].shape[0]
  File "/home/travis/build/pymc-devs/pymc3/pymc3/theanof.py", line 49, in floatX
    return X.astype(theano.config.floatX)
AttributeError: 'int' object has no attribute 'astype'
======================================================================
ERROR: test_advi_minibatch_shared (pymc3.tests.test_advi.TestADVI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_advi.py", line 221, in test_advi_minibatch_shared
    total_size=n, learning_rate=1e-1)
  File "/home/travis/minconda3/envs/testenv/lib/python3.5/site-packages/theano/configparser.py", line 115, in res
    return f(*args, **kwargs)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/variational/advi_minibatch.py", line 459, in advi_minibatch
    minibatch_tensors, total_size)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/variational/advi_minibatch.py", line 45, in _get_rvss
    s = floatX(total_size) / minibatch_tensors[0].shape[0]
  File "/home/travis/build/pymc-devs/pymc3/pymc3/theanof.py", line 49, in floatX
    return X.astype(theano.config.floatX)
AttributeError: 'int' object has no attribute 'astype'
``` 
Is this related to the other suggestions by @ferrine ?

twiecki committed:
ENH default to asarray if int or float is passed.

twiecki commented:
Thanks for the review.
